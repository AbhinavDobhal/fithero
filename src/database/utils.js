/* @flow */

import leftPad from 'left-pad';

export const getExerciseSchemaId = (day: string, exerciseKey: string) =>
  `${day}_${exerciseKey}`;

export const getExerciseSchemaIdFromSet = (setId: string) => {
  const parts = setId.split('_');
  return `${parts[0]}_${parts[1]}`;
};

export const getSetSchemaId = (
  day: string,
  exerciseKey: string,
  index: number
) => `${getExerciseSchemaId(day, exerciseKey)}_${leftPad(index, 3, '0')}`;

export const extractWorkoutKeyFromDatabase = (id: string) => id.split('_')[0];

export const extractExerciseKeyFromDatabase = (id: string) => id.split('_')[1];

export const extractSetIndexFromDatabase = (id: string) =>
  parseInt(id.split('_')[2], 10);
