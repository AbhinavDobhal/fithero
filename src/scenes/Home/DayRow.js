/* @flow */

import * as React from 'react';
import { StyleSheet, View } from 'react-native';
import { Text } from 'react-native-paper';

import DayItem from './DayItem';
import {
  dateToString,
  getCurrentWeek,
  getDatePrettyFormat,
  getToday,
  isSameDay,
} from '../../utils/date';

type Props = {
  onDaySelected: string => void,
  selected: string,
};

class DayRow extends React.Component<Props> {
  _onDaySelected = dateString => {
    this.props.onDaySelected(dateString);
  };

  _renderDays() {
    const days = getCurrentWeek(new Date());
    return days.map(d => {
      const isSelected = isSameDay(d, this.props.selected);
      const dateString = dateToString(d);
      return (
        <DayItem
          key={dateString}
          dateString={dateString}
          isSelected={isSelected}
          onDaySelected={this._onDaySelected}
        />
      );
    });
  }

  render() {
    return (
      <React.Fragment>
        <Text style={styles.title}>
          {getDatePrettyFormat(this.props.selected, getToday())}
        </Text>
        <View style={styles.row}>{this._renderDays()}</View>
      </React.Fragment>
    );
  }
}

const styles = StyleSheet.create({
  row: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: 8,
  },
  title: {
    fontSize: 15,
    paddingTop: 16,
    paddingBottom: 8,
    paddingHorizontal: 16,
  },
});

export default DayRow;
