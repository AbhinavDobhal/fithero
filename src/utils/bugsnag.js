/* @flow */

/**
 * Create a singleton instance of the bugsnag client so we don't have to duplicate our configuration
 * anywhere.
 *
 * https://docs.bugsnag.com/platforms/react-native/#basic-configuration
 */
import { Client } from 'bugsnag-react-native';

const client = new Client('7cb80c4b3678382f5880b407003b05b1');

module.exports = client;
