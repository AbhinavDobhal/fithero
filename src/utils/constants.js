/* @flow */

import { name as packageName } from '../../package.json';

export const Settings = {
  editSetsScreen: `${packageName}.preference.EDIT_SET_SCREEN`,
  defaultUnitSystem: `${packageName}.preference.DEFAULT_UNIT_SYSTEM`,
};
