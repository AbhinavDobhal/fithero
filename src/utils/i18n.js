/* @flow */

import I18n from 'react-native-i18n';
import { en, es } from './locales';

I18n.fallbacks = true;
I18n.translations = {
  en,
  es,
};

export default I18n;
