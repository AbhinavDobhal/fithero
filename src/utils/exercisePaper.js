/* @flow */

import type { ExerciseLog } from '../types';
import type { SetSchemaType } from '../database/types';
import { getSetSchemaId } from '../database/utils';

export const parseSummary = (
  exerciseSummary: string,
  day: string,
  exerciseKey: string
): ExerciseLog => {
  const sets: Array<SetSchemaType> = [];
  let comments = '';

  const lines = exerciseSummary
    .split('\n')
    .map(line => line.trim())
    .filter(line => line !== '');

  let setIndex = 1;
  lines.forEach((line, index) => {
    // We remove whitespaces too so we fix mistakes like "5x 30"
    const isSet = /\d+x\d+\.?\d*/.test(line.replace(/\s/g, ''));
    if (isSet) {
      const set = line.split('x');
      // We do not allow a set with 0 reps
      const reps = parseInt(set[0], 10);
      if (reps > 0) {
        sets.push({
          id: getSetSchemaId(day, exerciseKey, setIndex),
          reps,
          // Weight can be a float (reps cannot)
          weight: parseFloat(set[1]),
        });
        setIndex++;
      }
    } else if (index === lines.length - 1) {
      comments = line;
    }
  });

  return {
    sets,
    comments,
  };
};

export const generateSummary = (log: ExerciseLog) => {
  const sets = [];
  log.sets.forEach(set => {
    sets.push(`${set.reps}x${set.weight}`);
  });

  return `${sets.join(`\n`)}${log.comments ? `\n\n${log.comments}` : ''}`;
};
