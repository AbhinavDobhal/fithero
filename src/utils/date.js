/* @flow */

// import { NativeModules } from 'react-native';
import moment from 'moment';
import * as RNLocalize from 'react-native-localize';

// To import locales we want translations from
// import 'moment/locale/pl';
// import 'moment/locale/es';
// moment.locale(NativeModules.RNI18n.languages[0]);

// To change the starting week day to Monday
// Although locale do it for you, it can override the locale
// moment.updateLocale('en', {
//   week: {
//     dow: 1,
//   },
// });

import i18n from './i18n';
import regionDayMap from './regionDayMap';
import type { FirstDayOfTheWeekType } from '../redux/modules/settings';

// TODO this function won't work properly if we change the moment.locale
const getFirstDayOfTheWeek = (
  date: Date,
  firstDayOfTheWeek: FirstDayOfTheWeekType
) => {
  const now = moment(date);
  const firstDay = firstDayOfTheWeekToNumber(firstDayOfTheWeek);
  const firstWeekDay = now
    .clone()
    .weekday(firstDay)
    .startOf('day');
  if (!firstWeekDay.isAfter(now)) {
    return firstWeekDay;
  }
  return now
    .clone()
    .weekday(firstDay - 7)
    .startOf('day');
};

export const getCurrentWeek = (
  date: Date,
  firstDayOfTheWeek: FirstDayOfTheWeekType
) => {
  const start = getFirstDayOfTheWeek(date, firstDayOfTheWeek);
  const days = [start.clone().toDate()];

  for (let i = 1; i <= 6; i++) {
    days.push(
      start
        .clone()
        .add(i, 'day')
        .toDate()
    );
  }

  return days;
};

export const getFirstAndLastWeekday = (
  date: Date,
  firstDayOfTheWeek: FirstDayOfTheWeekType
) => {
  const now = moment(date);
  const firstDay = firstDayOfTheWeekToNumber(firstDayOfTheWeek);
  const start = getFirstDayOfTheWeek(date, firstDayOfTheWeek);
  const end = now
    .clone()
    .weekday(firstDay + 6)
    .startOf('day');

  return [start.toDate(), end.toDate()];
};

export const getShortDayInfo = (dateString: string) => {
  const momentDate = moment(dateString);
  return {
    date: momentDate.get('date'),
    day: momentDate.format('ddd'),
  };
};

export const getToday = () => moment().startOf('day');

export const isSameDay = (date: Date, today: string) =>
  moment(date).isSame(moment(today), 'day');

export const dateToWorkoutId = (date: Date | string) =>
  moment(date).format('YYYYMMDD');

export const dateToString = (date: Date) => moment(date).toISOString();

export const toDate = (dateString: string) => moment(dateString).toDate();

export const getDatePrettyFormat = (
  dateString: Date | string,
  today: string,
  short?: boolean = false
) => {
  const date = moment(dateString);
  const isToday = date.isSame(moment(today), 'day');

  if (isToday) {
    return `${i18n.t('today')}${!short ? `, ${date.format('MMMM D')}` : ''}`;
  }
  return date.format(`${!short ? 'dddd' : 'ddd'}, MMMM D`);
};

export const getDay = (day: string) =>
  moment(day)
    .startOf('day')
    .toISOString();

export const formatDate = (date: Date | string, format: string) =>
  moment(date).format(format);

export const getFirstAndLastMonthDay = (date: Date) => {
  const now = moment(date);
  const start = now.startOf('month');
  const end = now.clone().endOf('month');

  return [start.toDate(), end.toDate()];
};

export const getWeekStartByLocale = (): FirstDayOfTheWeekType => {
  const country = RNLocalize.getCountry();

  const day = regionDayMap[country];
  if (day === 0) {
    return 'sunday';
  } else if (day === 6) {
    return 'saturday';
  }
  return 'monday';
};

export const firstDayOfTheWeekToNumber = (
  firstDayOfTheWeek: FirstDayOfTheWeekType
) => {
  let firstDay = 1; // Monday
  if (firstDayOfTheWeek === 'sunday') {
    firstDay = 0;
  } else if (firstDayOfTheWeek === 'saturday') {
    firstDay = 6;
  }
  return firstDay;
};

export const getSafeTimezoneTime = (date: Date) => {
  return moment(date)
    .subtract(1, 'day')
    .toDate();
};
