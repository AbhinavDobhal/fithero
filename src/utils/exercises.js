/* @flow */

import i18n from './i18n';

export const extractExerciseKeyFromDatabase = (exerciseId: string) =>
  exerciseId.split('_')[1];

export const getExerciseName = (key: string) => i18n.t(`exercise__${key}`);

export const getExerciseCategoryName = (category: string) =>
  i18n.t(`exercise-category__${category}`);
