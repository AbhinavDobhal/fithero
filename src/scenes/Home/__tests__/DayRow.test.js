/* @flow */

import * as React from 'react';
import { render } from 'react-native-testing-library';
import moment from 'moment';

import DayRow from '../DayRow';
import { getCurrentWeek } from '../../../utils/date';
import type { FirstDayOfTheWeekType } from '../../../redux/modules/settings';

describe('DayRow', () => {
  const getRender = (day: string, firstDayOfTheWeek: FirstDayOfTheWeekType) => {
    const currentWeek = getCurrentWeek(
      moment(day).startOf('day'),
      firstDayOfTheWeek
    );
    return render(
      <DayRow
        currentWeek={currentWeek}
        workouts={{}}
        onDaySelected={jest.fn()}
        selected={day}
      />
    );
  };

  const { toJSON: mondayJSON } = getRender(
    '2019-03-18T00:00:00.000Z',
    'monday'
  );

  it('day row starts on Monday or Sunday', () => {
    const { toJSON: sundayJSON } = getRender(
      '2019-03-17T00:00:00.000Z',
      'sunday'
    );

    // $FlowFixMe
    expect(mondayJSON()).toMatchDiffSnapshot(sundayJSON(), {
      contextLines: 0,
    });
  });

  it('day row starts on Monday or Saturday', () => {
    const { toJSON: saturdayJSON } = getRender(
      '2019-03-16T00:00:00.000Z',
      'saturday'
    );

    // $FlowFixMe
    expect(mondayJSON()).toMatchDiffSnapshot(saturdayJSON(), {
      contextLines: 0,
    });
  });
});
