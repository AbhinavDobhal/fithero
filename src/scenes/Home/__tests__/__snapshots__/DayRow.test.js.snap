// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayRow day row starts on Monday or Saturday 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ --- --- @@
-     Monday, March 18
+     Saturday, March 16
@@ --- --- @@
-             18
+             16
@@ --- --- @@
-             Mon
+             Sat
@@ --- --- @@
-             19
+             17
@@ --- --- @@
-             Tue
+             Sun
@@ --- --- @@
-             20
+             18
@@ --- --- @@
-             Wed
+             Mon
@@ --- --- @@
-             21
+             19
@@ --- --- @@
-             Thu
+             Tue
@@ --- --- @@
-             22
+             20
@@ --- --- @@
-             Fri
+             Wed
@@ --- --- @@
-             23
+             21
@@ --- --- @@
-             Sat
+             Thu
@@ --- --- @@
-             24
+             22
@@ --- --- @@
-             Sun
+             Fri
`;

exports[`DayRow day row starts on Monday or Sunday 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ --- --- @@
-     Monday, March 18
+     Sunday, March 17
@@ --- --- @@
-             18
+             17
@@ --- --- @@
-             Mon
+             Sun
@@ --- --- @@
-             19
+             18
@@ --- --- @@
-             Tue
+             Mon
@@ --- --- @@
-             20
+             19
@@ --- --- @@
-             Wed
+             Tue
@@ --- --- @@
-             21
+             20
@@ --- --- @@
-             Thu
+             Wed
@@ --- --- @@
-             22
+             21
@@ --- --- @@
-             Fri
+             Thu
@@ --- --- @@
-             23
+             22
@@ --- --- @@
-             Sat
+             Fri
@@ --- --- @@
-             24
+             23
@@ --- --- @@
-             Sun
+             Sat
`;
